nutanix_user_key_v2 update tries to recreate the key object

# Nutanix Cluster Information

Prism Central Version: pc.7.5.0.5

AOS Version: 7.5.0.5

## Terraform Version

Terraform: v1.14.2

Provider: registry.terraform.io/nutanix/nutanix v2.4.0

## Affected Resources

"nutanix_user_key_v2"

## Terraform Configuration Files

### Create

```terraform
# Generate suffix
resource "random_string" "suffix" {
  length  = 6
  upper   = false
  special = false
}

# Create a service account
resource "nutanix_users_v2" "service_account" {
  username    = "service_account_${random_string.suffix.result}"
  description = "service account for object storage"
  email_id    = "service_account_object_storage@somedomain.com"
  user_type   = "SERVICE_ACCOUNT"
}

# Create a object storage key
resource "nutanix_user_key_v2" "object_storage_key" {
  depends_on  = [nutanix_users_v2.service_account]
  user_ext_id = nutanix_users_v2.service_account.ext_id
  name        = "key_for_object_storage_${random_string.suffix.result}"
  key_type    = "OBJECT_KEY"
}

# Create an api key
resource "nutanix_user_key_v2" "api_key" {
  depends_on  = [nutanix_users_v2.service_account]
  user_ext_id = nutanix_users_v2.service_account.ext_id
  name        = "key_for_api_${random_string.suffix.result}"
  key_type    = "API_KEY"
}

```

### Update

```terraform
variable "object_key_ext_id" {
  description = "The ext_id for an existing object key"
  type        = string
}

variable "api_key_ext_id" {
  description = "The ext_id for an existing api key"
  type        = string
}

variable "service_account_ext_id" {
  description = "The ext_id for an existing user"
  type        = string
}

import {
  id = format("%s/%s", var.service_account_ext_id, var.api_key_ext_id)
  to = nutanix_user_key_v2.api_key
}

import {
  id = format("%s/%s", var.service_account_ext_id, var.object_key_ext_id)
  to = nutanix_user_key_v2.object_storage_key
}


resource "nutanix_user_key_v2" "api_key" {
  user_ext_id = var.service_account_ext_id
  name        = "unknown"
  key_type    = "API_KEY"
  description = "updated api_key description"

  lifecycle {
    ignore_changes = [name, key_type]
  }
}

resource "nutanix_user_key_v2" "object_storage_key" {
  user_ext_id = var.service_account_ext_id
  name        = "unknown"
  key_type    = "OBJECT_KEY"
  description = "updated object_storage_key description"

  lifecycle {
    ignore_changes = [name, key_type]
  }
}
```

## Update Debug Output

```bash

[ERROR] vertex "nutanix_user_key_v2.api_key" error: error while creating User Key: {"$dataItemDiscriminator":"iam.v4.error.ErrorResponse","$objectType":"iam.v4.authn.CreateKeyApiResponse","$reserved":{"$fv":"v4.r1.b2"},"data":{"$errorItemDiscriminator":"List\u003ciam.v4.error.AppMessage\u003e","$objectType":"iam.v4.error.ErrorResponse","$reserved":{"$fv":"v4.r1.b2"},"error":[{"$objectType":"iam.v4.error.AppMessage","$reserved":{"$fv":"v4.r1.b2"},"argumentsMap":{"cause":"unable to parse request body","operation":"create key"},"code":"IAM-21206","errorGroup":"IAM-AUTHN-APIKEY-VAL-ERR","locale":"en_US","message":"Failed to create key as unable to parse request body.","severity":"ERROR"}]}}
```

### CRUDI Operations

#### Create ⚠️

secret_key field is empty string in the response so the key is not usable which is a known issue.

#### Read ✔️

#### Update ❌

It seems like the update tries to recreate the key object but sends a create request body containing invalid fields.

#### Delete ✔️

#### Import ✔️

## Steps To Reproduce

1. Create separate main.tf files with the tf configs above.
2. Issue `terraform apply` on the **Create** tf file
3. Issue `terraform apply` on the **Update** tf file

--- a/nutanix/services/iamv2/resource_nutanix_user_key_v2.go
+++ b/nutanix/services/iamv2/resource_nutanix_user_key_v2.go
@@ -63,7 +63,7 @@ func ResourceNutanixUserKeyV2() *schema.Resource {
 				Type:         schema.TypeString,
 				Required:     true,
 				ValidateFunc: validation.StringInSlice([]string{"API_KEY", "OBJECT_KEY"}, false),
-				ForceNew: true,
+				ForceNew:     true,
 			},
 			"created_time": {
 				Type:     schema.TypeString,

--- a/nutanix/services/iamv2/resource_nutanix_user_key_v2.go
+++ b/nutanix/services/iamv2/resource_nutanix_user_key_v2.go
@@ -51,16 +51,19 @@ func ResourceNutanixUserKeyV2() *schema.Resource {
 			"name": {
 				Type:     schema.TypeString,
 				Required: true,
+				ForceNew: true,
 			},
 			"description": {
 				Type:     schema.TypeString,
 				Optional: true,
 				Computed: true,
+				ForceNew: true,
 			},
 			"key_type": {
 				Type:         schema.TypeString,
 				Required:     true,
 				ValidateFunc: validation.StringInSlice([]string{"API_KEY", "OBJECT_KEY"}, false),
+				ForceNew: true,
 			},
 			"created_time": {
 				Type:     schema.TypeString,
@@ -82,6 +85,7 @@ func ResourceNutanixUserKeyV2() *schema.Resource {
 				Type:     schema.TypeString,
 				Optional: true,
 				Computed: true,
+				ForceNew: true,
 			},
 			"status": {
 				Type:         schema.TypeString,
@@ -108,6 +112,7 @@ func ResourceNutanixUserKeyV2() *schema.Resource {
 				Type:     schema.TypeString,
 				Optional: true,
 				Computed: true,
+				ForceNew: true,
 			},
 			"key_details": {
 				Type:     schema.TypeList,
@@ -327,7 +332,7 @@ func flattenKeyDetails(oneOfKeyKeyDetails *import1.OneOfKeyKeyDetails) interface
 }
 
 func resourceNutanixUserKeyV2Update(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
-	return resourceNutanixUserKeyV2Create(ctx, d, m)
+	return nil
 }
 
 func resourceNutanixUserKeyV2Delete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {

--- a/nutanix/services/objectsv2/resource_nutanix_object_store_v2.go
+++ b/nutanix/services/objectsv2/resource_nutanix_object_store_v2.go
@@ -63,8 +63,9 @@ func ResourceNutanixObjectStoresV2() *schema.Resource {
 				},
 			},
 			"name": {
-				Type:     schema.TypeString,
-				Required: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringLenBetween(1, 16),
 			},
 			"description": {
 				Type:     schema.TypeString,

--- a/nutanix/services/objectsv2/resource_nutanix_object_store_v2.go
+++ b/nutanix/services/objectsv2/resource_nutanix_object_store_v2.go
@@ -65,7 +65,7 @@ func ResourceNutanixObjectStoresV2() *schema.Resource {
 			"name": {
 				Type:         schema.TypeString,
 				Required:     true,
-				ValidateFunc: validation.StringLenBetween(1, 50),
+				ValidateFunc: validation.StringLenBetween(1, 16),
 			},
 			"description": {
 				Type:     schema.TypeString,

--- a/nutanix/services/objectsv2/resource_nutanix_object_store_v2.go
+++ b/nutanix/services/objectsv2/resource_nutanix_object_store_v2.go
@@ -65,7 +65,7 @@ func ResourceNutanixObjectStoresV2() *schema.Resource {
 			"name": {
 				Type:         schema.TypeString,
 				Required:     true,
-				ValidateFunc: validation.StringLenBetween(1, 16),
+				ValidateFunc: validation.StringLenBetween(1, 50),
 			},
 			"description": {
 				Type:     schema.TypeString,

--- a/website/docs/r/object_store_v2.html.markdown
+++ b/website/docs/r/object_store_v2.html.markdown
@@ -13,7 +13,7 @@ Run the prechecks, create and start the deployment of an Object store on Prism C
 > ⚠️ **Warning:** Before deleting the Object Store, make sure to delete all buckets inside it manually.
 > Currently, the Terraform provider does not support the Delete Bucket API.
 
-> ⚠️ **Warning:** The Object Store **update** operation does **not** allow modification of any configuration parameters (including `name` and `description`). Apart from retrying a failed deployment, no other changes can be performed through update.
+> ⚠️ **Warning:** The Object Store **update** operation does **not** allow modification of any configuration parameters (including `name` and `description`). As per the API design, the update operation is supported exclusively for retrying a failed deployment. No other changes can be performed through update.
 > It should be used when the Object Store is in the `OBJECT_STORE_DEPLOYMENT_FAILED` state.
 > Triggering an update in this state will attempt to resume the deployment process.
 

--- a/website/docs/r/object_store_v2.html.markdown
+++ b/website/docs/r/object_store_v2.html.markdown
@@ -13,7 +13,7 @@ Run the prechecks, create and start the deployment of an Object store on Prism C
 > ⚠️ **Warning:** Before deleting the Object Store, make sure to delete all buckets inside it manually.
 > Currently, the Terraform provider does not support the Delete Bucket API.
 
-> ⚠️ **Warning:** The Object Store **update** operation is intended **only** to resume a failed deployment.
+> ⚠️ **Warning:** The Object Store **update** operation does **not** allow modification of any configuration parameters (including `name` and `description`). Apart from retrying a failed deployment, no other changes can be performed through update.
 > It should be used when the Object Store is in the `OBJECT_STORE_DEPLOYMENT_FAILED` state.
 > Triggering an update in this state will attempt to resume the deployment process.
 

--- a/website/docs/r/user_key_v2.html.markdown
+++ b/website/docs/r/user_key_v2.html.markdown
@@ -23,6 +23,26 @@ resource "nutanix_user_key_v2" "create_key" {
 }
 ```
 
+## Lifecycle Behavior
+
+~> Important: The nutanix_user_key_v2 resource does not support in-place updates.
+
+Changes to the following arguments will force the resource to be replaced:
+
+- name
+
+- description
+
+- key_type
+
+- expiry_time
+
+- assigned_to
+
+When any of these arguments are modified, Terraform will destroy the existing user key and create a new one. This results in a new key being generated.
+
+~> Note: Replacing the resource invalidates the previously generated key. Ensure that any dependent systems are updated before applying the changes.
+
 ## Argument Reference
 
 The following arguments are supported:

